### DISCUSSION TITLE
quickly parameter-dependent assembly

### DISCUSSION THREAD

**Discussion by AnyWayOne** [kinnala/scikit-fem#732] [Q&A] [answered]
I wonder is it possible to implement quickly parameter-dependent assembly, e.g. the problem $-\triangledown(k(p)\triangledown(u))=1$, where $k(p)$ is a  function related to the parameter $p$.
A normal implementation is:
##############
mesh = Mesh...
Vh = Basis...
@BilinearForm
def a(u,v,w);
     return dot(w['kp']*grad(u), grad(v))
for k_p_i in k_p:
     A_i = a.assemble(Vh, kp = Vh.interpolate(k_p_i))
####################
$A_i \in R^{n \times n}$
By using a loop, it executes n_p (the number of the samplr of the parameter $p$) assemblys, but it's too time consuming for a large $n_p$.
Is it possible to implement above loop efficiently at once and to obtain a tensor $A \in R^{n \times n \times n_p}$?
Thanks for any answer.

**Comment by kinnala** [comment, marked as answer]
There is no command to do them all at once. In principle it would be possible to have a "TrilinearForm" to assemble 3-tensors and then do multiple contractions to 2-tensors. However, I think the limitation is that there is no efficient data structure (similar to scipy.sparse) available for such sparse 3-tensors that would be suitable for FE assembly (sums multiple values during init).

One thing you could try is using Dask (and dask.bag) to parallelize the for-loop. Those computations do not depend on one another so if you have n_p computers, it should be pretty quick.

**Reply by kinnala** [reply]
I tried doing Dask approach but it seems that you need to move around quite big chunks of data in comparison to the fast speed of a single assembly for such a simple form. So I think that might not help you out after all.

**Reply by AnyWayOne** [reply]
Thank your for your reply. I also tested dask, it has little effect (although it can save some time). I am trying to modify the code to add an extra dimension to associate the parameter dimension. It seems to not a simple matter. o(╥﹏╥)o

**Reply by kinnala** [reply]
I think this code is technically doing the correct thing:
```python
from typing import Optional, Tuple, Any

import numpy as np
from numpy import ndarray

from skfem import *
from skfem.assembly.form.form import Form
from skfem.assembly.form.form import FormExtraParams


class TrilinearForm(Form):

    def _assemble(self,
                  ubasis: Basis,
                  vbasis: Optional[Basis] = None,
                  wbasis: Optional[Basis] = None,
                  **kwargs) -> Tuple[ndarray,
                                     ndarray,
                                     ndarray,
                                     ndarray,
                                     Tuple[int, int, int]]:

        if vbasis is None:
            vbasis = ubasis
        if wbasis is None:
            wbasis = ubasis

        nt = ubasis.nelems
        dx = ubasis.dx
        wdict = FormExtraParams({
            **ubasis.default_parameters(),
            **self.dictify(kwargs, ubasis),
        })

        # initialize COO data structures
        sz = (ubasis.Nbfun, vbasis.Nbfun, wbasis.Nbfun, nt)
        data = np.zeros(sz, dtype=self.dtype)
        rows = np.zeros(sz, dtype=np.int64)
        cols = np.zeros(sz, dtype=np.int64)
        mats = np.zeros(sz, dtype=np.int64)

        # loop over the indices of local stiffness matrix
        for k in range(ubasis.Nbfun):
            for j in range(vbasis.Nbfun):
                for i in range(wbasis.Nbfun):
                    mats[k, j, i] = wbasis.element_dofs[i]
                    rows[k, j, i] = vbasis.element_dofs[j]
                    cols[k, j, i] = ubasis.element_dofs[k]
                    data[k, j, i] = self._kernel(
                        ubasis.basis[k],
                        vbasis.basis[j],
                        wbasis.basis[i],
                        wdict,
                        dx,
                    )

        return data, mats, rows, cols, (wbasis.N, vbasis.N, ubasis.N)

    def assemble(self, *args, **kwargs) -> Any:
        return self._assemble(*args, **kwargs)

    def _kernel(self, u, v, w, params, dx):
        return np.sum(self.form(*u, *v, *w, params) * dx, axis=1)



m = MeshTri().refined(3)
basis = Basis(m, ElementTriP1())
from skfem.helpers import *
out = TrilinearForm(lambda u, v, w, p: dot(w * grad(u), grad(v))).assemble(basis)
```
However, now your issue is that how to use `out` because that is not `scipy.sparse.csr_matrix` or anything nice like that. It is a sparse COO 3-tensor with duplicate entries. You would need to apply something like https://pytorch.org/docs/stable/sparse.html to use it efficiently. And I have no idea how to sum the duplicate entries efficiently since this is usually done by `scipy.sparse.coo_matrix` constructor which is implemented in a low-level language.

**Reply by kinnala** [reply]
In theory if you had your `k_p` in a matrix and did correct tensor operations between `k_p` and `out` you would arrive at the result.

**Reply by AnyWayOne** [reply]
Amazing Job. I think the code is efficient enough now. After obtaining the 3-tensor, I just contract it to a 2-tensor by a simple computation.

**Reply by kinnala** [reply]
If you are able to figure out how to do the contraction efficiently I'm interested in hearing about that.

**Reply by kinnala** [reply]
Actually this library seems promising: https://sparse.pydata.org/en/stable/

**Reply by AnyWayOne** [reply]
For my computation, it needs an expectation computation, $E[rvs*A]$ (or very similar to this), where $rvs$ is a random variable or random field, $A \in R^{n \times n \times n_p}$. I only use the numpy.einsum to compute in the parameter dimension, which can be very efficient.

**Reply by kinnala** [reply]
Numpy supports only dense arrays. Is that not a problem for you? There are many nonzero values in the resulting 3-tensor.

I might add this `TrilinearForm` into the package but I'm not sure what kind of format it should output. I'm now successfully constructing `sparse.COO` and learning how to use this `sparse` package.

**Reply by AnyWayOne** [reply]
I strongly suggest adding the code into skfem, because it will be very efficient and convenient for parameter analysis, sensitivity analysis and uncertainty quantification, etc.

**Reply by AnyWayOne** [reply]
Currently, dense matrices are acceptable to my problems. It will become sparse again after a simple computation. It would be best if a sparse form can be directly obtained.

**Reply by kinnala** [reply]
Adding to the above code here is a snippet that I used:
```python
def asm():
    m = MeshTri().refined(7)
    basis = Basis(m, ElementTriP1())
    from skfem.helpers import dot, grad
    out = TrilinearForm(lambda u, v, w, p: dot(w * grad(u), grad(v))).assemble(basis)
    import sparse
    arr = sparse.COO(np.array([out[1].flatten(), out[2].flatten(), out[3].flatten()]), out[0].flatten(), out[4], has_duplicates=True)
    k_p = np.random.rand(100, basis.N)
    res = sparse.tensordot(arr, k_p.T, axes=1, return_type=sparse.COO)

    A = []
    for mat in res.T:
        A.append(mat.to_scipy_sparse().tocsr())
    return A
```
It seems to run pretty quickly. However, I don't want to introduce an additional hard dependency to `sparse` because it depends on `numba` which is a rather heavy dependency and we try to keep it minimal. Therefore I think the best approach is to return something which is directly compatible with the constructor of `sparse.COO`.

**Reply by AnyWayOne** [reply]
It seems to be quick. I'll test it later on larger problems.
For the output of the sparse 3-tensor, generally there is no guarantee that the matrix with the third dimension is sparse, I think a possible format is to slice in the third dimension, i.e., A(:,:,i) is constructed as a sparse matrix, but A(i,:,:) or A(:,i,:) may be not a sparse matrix. This is just an extension of the concept of a two-dimensional sparse matrix. I have no idea about the general purpose of 3-tensors, this is only from the point of my problems.

**Reply by kinnala** [reply]
Continuing to track this is #733.
